langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/DefaultMcpClient.java

```diff
@@ -0,0 +1,104 @@
+package dev.langchain4j.mcp.client;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import dev.langchain4j.agent.tool.ToolSpecification;
+import dev.langchain4j.mcp.client.protocol.InitializeParams;
+import dev.langchain4j.mcp.client.protocol.McpCallToolRequest;
+import dev.langchain4j.mcp.client.protocol.McpInitializeRequest;
+import dev.langchain4j.mcp.client.protocol.McpListToolsRequest;
+import dev.langchain4j.mcp.client.transport.McpTransport;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO: currently we request a new list of tools every time, perhaps we can add some configurable caching
+public class DefaultMcpClient implements McpClient {
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultMcpClient.class);
+    private final AtomicLong idGenerator = new AtomicLong(0);
+    private final McpTransport transport;
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    public DefaultMcpClient(Builder builder) {
+        transport = builder.transport;
+        transport.start();
+        McpInitializeRequest request = new McpInitializeRequest(idGenerator.getAndIncrement());
+        InitializeParams params = createInitializeParams();
+        request.setParams(params);
+        JsonNode capabilities = transport.initialize(request);
+        log.info("MCP server capabilities: {}", capabilities.get("result"));
+    }
+
+    private InitializeParams createInitializeParams() {
+        InitializeParams params = new InitializeParams();
+        params.setProtocolVersion("2024-11-05"); // TODO: what is the correct value?
+
+        InitializeParams.ClientInfo clientInfo = new InitializeParams.ClientInfo();
+        clientInfo.setName("langchain4j"); // TODO: this probably should be configurable
+        clientInfo.setVersion("0.1"); // TODO: this probably should be configurable
+        params.setClientInfo(clientInfo);
+
+        InitializeParams.Capabilities capabilities = new InitializeParams.Capabilities();
+        InitializeParams.Capabilities.Roots roots = new InitializeParams.Capabilities.Roots();
+        roots.setListChanged(false); // TODO: listChanged is not supported yet
```

Comment:
Yes, and I will implement that soon

Reply:
One problem is that I don't know of any MCP server that actually does this - do you know any? If there's none, I don't have any way to test the implementation and it may not make much sense to do (yet)

Reply:
I don't know any, but since adding/changing tools usually require writing code, I am not sure it is very useful. I can imagine new versions of MCP servers having more tools, but that will require restarting the connection anyway? Perhaps notifications are more useful for resources/data then for tools

Reply:
A more advanced server could potentially serve tools that are dynamically added and removed and it wouldn't require restarting,  but I don't know if that's going to become a normal thing. Maybe we could leave this for now and see how stuff pans out later.

Reply:
Makes sense